dune/iga/reconstructedgridhandler.hh

```diff
@@ -0,0 +1,285 @@
+    inline static auto get(const Dune::FieldVector<T, 2>& t) { return t[1]; };
+  };
+}  // namespace mapbox::util
+
+namespace Dune::IGA {
+
+  template <int worldDim>
+  struct GridBoundarySegment : Dune::BoundarySegment<2, worldDim, double> {
+    // Types
+    using LocalPoint  = Dune::FieldVector<double, 2>;
+    using GlobalPoint = Dune::FieldVector<double, worldDim>;
+
+    // Constructor
+    explicit GridBoundarySegment(Boundary& _boundary, auto _transferToGlobal)
+        : boundary(_boundary), transferToGlobal{_transferToGlobal} {}
+
+    Dune::FieldVector<double, worldDim> operator()(const Dune::FieldVector<double, 1>& local) const override {
+      // u muss auf die Domäne gemappt werden, local[0] ist zwischen 0 und 1
+      double u_local = local[0];
+      double u       = Utilities::map<double>(u_local, 0.0, 1.0, boundary.domain[0], boundary.domain[1]);
+      auto res       = boundary.nurbsGeometry(u);
+
+      // Transfer to global Surface Coordinates
+      return (*transferToGlobal)({res[0], res[1]});
+    }
+
+    Boundary boundary;
+    std::shared_ptr<std::function<GlobalPoint(LocalPoint)>> transferToGlobal;
+  };
+
+  template <int worldDim>
+  class ReconstructedGridHandler {
+   public:
+    using Grid     = Dune::UGGrid<worldDim>;
+    using GridView = Grid::LeafGridView;
+
+    // Abkürzungen
+    using BoundaryVector = std::vector<Boundary>;
+    using LocalPoint     = Dune::FieldVector<double, 2>;
+    using GlobalPoint    = Dune::FieldVector<double, worldDim>;
+
+    using LocalPointVector  = std::vector<LocalPoint>;
+    using GlobalPointVector = std::vector<GlobalPoint>;
+
+    // Grid
+    std::unique_ptr<Grid> grid;
+
+   private:
+    BoundaryVector boundaries;
+    std::shared_ptr<std::function<GlobalPoint(LocalPoint)>> transferToGlobal;
+    std::array<int, 2> gridDegree;
+
+   public:
+    // Construct with igaGrid and boundaries
+    ReconstructedGridHandler(BoundaryVector& _boundaries, auto _transferToGlobal, std::array<int, 2> _gridDegree)
+        : boundaries(_boundaries), transferToGlobal(_transferToGlobal), gridDegree(_gridDegree) {
+      reconstructTrimmedElement();
+    }
+
+   private:
+    void reconstructTrimmedElement() {
+      // Getting global Parameters
+      auto parameters = Utilities::getParameters();
+
+      if (parameters.preSample > 0) splitBoundaries();
+
+      LocalPointVector vertices;
+      for (auto& boundary : boundaries) {
+        vertices.push_back(boundary.controlPoints[0]);
+      }
+
+      GlobalPointVector globalVertices;
+      for (auto& point : vertices) {
+        globalVertices.push_back((*transferToGlobal)({point[0], point[1]}));
+      }
+
+      // Grid Factory
+      Dune::GridFactory<Grid> gridFactory;
+
+      for (auto& vertex : globalVertices)
+        gridFactory.insertVertex(vertex);
+
+      // Make Grid Template
+      auto meshTemplate = createOriginalMeshTemplate<unsigned int>();
+
+      int n_ele = (int)meshTemplate.size() / 3;
```

Comment:
`3` or `3 * i + 2` are not clear from the context, what these magic numbers mean here.
You should give them a name.

Reply:
i.e. `const int verticesPerTriangle =3`